include/pmacc/memory/buffers/Exchange.hpp

```diff
@@ -41,19 +52,218 @@ namespace pmacc
+            : hostBuffer(nullptr)
+            , deviceDoubleBuffer(nullptr)
+            , exchange(extype)
+            , communicationTag(commTag)
+        {
+            using DeviceBuffer = DeviceBuffer<TYPE, DIM>;
+            deviceBuffer = std::make_unique<DeviceBuffer>(exchangeDataSpace, sizeOnDevice);
+            //  this->deviceBuffer = new DeviceBuffer<TYPE, DIM > (exchangeDataSpace, sizeOnDevice,true);
+            if constexpr(DIM > DIM1)
+            {
+                /*create double buffer on gpu for faster memory transfers*/
+                deviceDoubleBuffer = std::make_unique<DeviceBuffer>(exchangeDataSpace, false, true);
+            }
+
+            if(!Environment<>::get().isMpiDirectEnabled())
+            {
+                using HostBuffer = HostBuffer<TYPE, DIM>;
+                hostBuffer = std::make_unique<HostBuffer>(exchangeDataSpace);
+            }
+        }
+
         /**
-         * Returns the exchange buffer on the device.
-         *
-         * @return Exchange buffer on device
+         * specifies in returned DataSpace which dimensions exchange data
+         * @param exchange the exchange mask
+         * @return DIM1 DataSpace of size 3 where 1 means exchange, 0 means no exchange
          */
-        virtual DeviceBuffer<TYPE, DIM>& getDeviceBuffer() = 0;
+        DataSpace<DIM> exchangeTypeToDim(uint32_t exchange) const
+        {
+            DataSpace<DIM> result;
+
+            Mask exchangeMask(exchange);
+
+            if(exchangeMask.containsExchangeType(LEFT) || exchangeMask.containsExchangeType(RIGHT))
+                result[0] = 1;
+
+            if constexpr(DIM > DIM1)
+                if(exchangeMask.containsExchangeType(TOP) || exchangeMask.containsExchangeType(BOTTOM))
+                    result[1] = 1;
+
+            if constexpr(DIM > DIM2)
+                if(exchangeMask.containsExchangeType(FRONT) || exchangeMask.containsExchangeType(BACK))
+                    result[2] = 1;
+
+            return result;
+        }
+
+        virtual ~Exchange() = default;
+
+        DataSpace<DIM> exchangeTypeToOffset(
+            uint32_t exchange,
+            GridLayout<DIM>& memoryLayout,
+            DataSpace<DIM> guardingCells,
+            uint32_t area) const
+        {
+            DataSpace<DIM> size = memoryLayout.getDataSpace();
+            DataSpace<DIM> border = memoryLayout.getGuard();
+            Mask mask(exchange);
+            DataSpace<DIM> tmp_offset;
+            if constexpr(DIM >= DIM1)
+            {
+                if(mask.containsExchangeType(RIGHT))
+                {
+                    tmp_offset[0] = size[0] - border[0] - guardingCells[0];
+                    if(area == GUARD)
+                    {
+                        tmp_offset[0] += guardingCells[0];
+                    }
+                    /* std::cout<<"offset="<<tmp_offset[0]<<"border"<<border[0]<<std::endl;*/
+                }
+                else
```

Comment:
Can it happen, that a DataSpace contains both left and right exchange types? 
If yes, this would not be covered by this if/else branch.

Reply:
No, it can only be LEFT or RIGHT.

I will remove the std our in the  line above!